runtime/interpreter/interpreter.go

```diff
@@ -3981,17 +3990,15 @@ func (interpreter *Interpreter) authAccountLinkFunction(addressValue AddressValu
 				newCapabilityDomain,
 				newCapabilityIdentifier,
 			) {
-				return NilValue{}
+				return NewNilValue(invocation.Interpreter)
 			}
 
 			// Write new value
 
 			borrowStaticType := ConvertSemaToStaticType(borrowType)
 
-			linkValue := LinkValue{
-				TargetPath: targetPath,
-				Type:       borrowStaticType,
-			}
+			// Unmetered here because interpreter.writeStored(...) meters it
+			linkValue := NewUnmeteredLinkValue(targetPath, borrowStaticType)
```

Comment:
IINM `writeStored` doesn't meter upon storing. It only meter when a value is read-back from storage (decoding).

Reply:
That's true. `writeStored` here eventually calls `interpreter.ValidateAtreeValue`, which eventually calls `decodeLink`.

How's this?

 	// This is unmetered because this value won't be exposed to the user. It's internal.
	// The value returned to the user is metered because it goes through `decodeLink(...)`,
	// which is called by `writeStored(...)` below, after going through `ValidateAtreeValue(...)`.

Reply:
The `ValidateAtreeValue()` is hidden behind the`atreeValueValidationEnabled` flag, and is not always enabled. e.g: it is not enabled on mainnet due to performance reasons. 
On the other hand, at the time of decoding, the decoder creates another `LinkedValue` instance, so there are in fact two values in memory. So it's ok/correct to meter it here, regardless of whether it was metered again at atree validation (`decodeLink()`)

Reply:
https://github.com/onflow/cadence/pull/1512